### ISSUE TITLE
[Mono 5.8] InvalidCastException when building ASP MVC 5 project with RazorGenerator 

### ISSUE THREAD

**Issue by jtorres** [mono/mono#12955]
## Steps to Reproduce

1. Run `msbuild /r` and then `msbuild` in attached project (any project with RazorGenerator.Core nuget package is ok) - using VSfMac build also fails after restoring nuget packafges

[ASPMVC.zip](https://github.com/mono/mono/files/2855199/ASPMVC.zip)

### Current Behavior

`msbuild` throws InvalidCastException

### Expected Behavior

Should build successfully 

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

## Additional info

I tested several Mono MDK versions and everything works using 5.4.* and prior versions, and does not when using 5.8.* on.

This does not happen in VIsual Studio for Windows.

Also tested RazorGenerator.Core previous versions (from 2.0 to 2.5) and fails (using Mono 5.8 on)

Checked recent changes in `System.Web.Configuration` and `System.Configuration` in Mono repository since stacktrace states those are the classes where the Invalid Cast Exception is taking place but couldn't find any related change that might cause this bug.

Developer Community issue [link](https://developercommunity.visualstudio.com/content/problem/392664/invalidcastexception-when-building-asp-mvc-5-proje.html)

**It work using Mono 5.4 and before and fails using Mono 5.8 on.**

## Stacktrace

```
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : InvalidCastException: Specified cast is not valid.
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Web.Configuration.WebConfigurationHost.MapPathFromMapper (System.String virtualPath) [0x0001b] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web.Configuration_2.0/WebConfigurationHost.cs:275 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Web.Configuration.WebConfigurationHost.MapPath (System.String virtualPath) [0x00024] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web.Configuration_2.0/WebConfigurationHost.cs:232 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Web.Configuration.WebConfigurationHost.GetStreamName (System.String configPath) [0x00061] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web.Configuration_2.0/WebConfigurationHost.cs:148 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Configuration.Configuration.Init (System.Configuration.Internal.IConfigSystem system, System.String configPath, System.Configuration.Configuration parent) [0x0000e] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Configuration/System.Configuration/Configuration.cs:127 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Configuration.Configuration..ctor (System.Configuration.InternalConfigurationSystem system, System.String locationSubPath) [0x00056] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Configuration/System.Configuration/Configuration.cs:96 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Configuration.Configuration..ctor (System.Configuration.InternalConfigurationSystem system, System.String locationSubPath) [0x00038] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Configuration/System.Configuration/Configuration.cs:91 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Configuration.Configuration..ctor (System.Configuration.InternalConfigurationSystem system, System.String locationSubPath) [0x00038] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Configuration/System.Configuration/Configuration.cs:91 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Configuration.InternalConfigurationFactory.Create (System.Type typeConfigHost, System.Object[] hostInitConfigurationParams) [0x0000d] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Configuration/System.Configuration/InternalConfigurationFactory.cs:41 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at System.Web.Configuration.WebConfigurationManager.OpenMappedWebConfiguration (System.Web.Configuration.WebConfigurationFileMap fileMap, System.String path) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/System.Web/System.Web.Configuration_2.0/WebConfigurationManager.cs:275 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.Core.MvcWebConfigTransformer.Initialize (RazorGenerator.Core.RazorHost razorHost, System.Collections.Generic.IDictionary`2[TKey,TValue] directives) [0x0009a] in <a1d5ff075ac645ee99b1f76e144133f0>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.Core.AggregateCodeTransformer.Initialize (RazorGenerator.Core.RazorHost razorHost, System.Collections.Generic.IDictionary`2[TKey,TValue] directives) [0x0001c] in <a1d5ff075ac645ee99b1f76e144133f0>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.Core.MvcViewTransformer.Initialize (RazorGenerator.Core.RazorHost razorHost, System.Collections.Generic.IDictionary`2[TKey,TValue] directives) [0x00000] in <a1d5ff075ac645ee99b1f76e144133f0>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.Core.RazorCodeTransformerBase.RazorGenerator.Core.IRazorCodeTransformer.Initialize (RazorGenerator.Core.IRazorHost razorHost, System.Collections.Generic.IDictionary`2[TKey,TValue] directives) [0x00000] in <a1d5ff075ac645ee99b1f76e144133f0>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.Core.RazorHost.GenerateCode () [0x00000] in <a1d5ff075ac645ee99b1f76e144133f0>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : at RazorGenerator.MsBuild.RazorCodeGen.ExecuteCore () [0x00167] in <4a7bbe6caf2141718ca4797949d78361>:0 
/Users/ben/Dashboard/Dashboard/packages/RazorGenerator.MsBuild.2.5.0/build/RazorGenerator.MsBuild.targets(48,9): error : 
1 Warning(s)
1 Error(s)
```

**Comment by benbek** [comment]
Hey there, I'm the OP in the Developer Community.
I'll be happy to supply any information needed in hope for a quick turnaround.

Thanks, @jtorres!

**Comment by slluis** [comment]
@marek-safar this is a regression. Is there any chance this can be fixed before 6.0?

**Comment by marek-safar** [comment]
@slluis we'll try to land it to 2018-08
